WordPress/gutenberg
# Support drag and drop in horizontal block lists and improve drop zone detection

## Description
Fixes #22662

This PR adds support for drag and drop in horizontal block lists (buttons, social links, navigation) and also improves how we detect where the user intends to drop a block.

Creating a draft PR for now as this needs tests, but it'd be good to get some feedback.

### How does this work

Previously, block drag and drop worked by using the `y` position of the dragged item, and comparing that to the `y` position of the blocks being dragged over. When the `y` position was less than halfway over a block, the block would be dropped before, and when more than halfway dropped after.

For horizontal block lists there are more challenges, this system can't be translated to the `x` axis because horizontal block lists tend to wrap onto multiple lines (for example, the buttons block).

This PR instead compares the position of the dragged item to the relevant edges of blocks. The top and bottom edges for vertical blocks lists and the left and right edges for horizontal. It picks the edge of the block that's nearest to the dragged item and uses that to determine where the item should be dropped.

### Alternatives

I considered a few different alternatives and also tried a few different ways to calculate the drop position. One idea was drawing invisible block drop zones over the block list that can more concretely act as a drop target. This didn't work too badly, but I'd consider it a bigger change and it would also have issues if something has a higher z-index than the drop zone.

### Remaining issues

There are still some areas that can be improved:
- Because of the way the block lists are used as drop zones, it can be a little difficult to drag to the start of an InnerBlocks list when there's not much margin. That seems to be because the cursor has to be within the block list for it to be considered a valid target.
- The indication of the drop zone area could still be improved quite a bit. I plan to work on this in a separate PR.

## How has this been tested?
1. Add a mixture of blocks preferably using normal blocks and blocks that have inner blocks (e.g buttons, media text, columns, group)
2. Try dragging blocks around to see the general experience
3. Try dragging buttons
4. Observe that dragging in horizontal block lists now works correctly.

## Types of changes

Bug fix (non-breaking change which fixes an issue)



## Checklist:
